core/src/main/scala/org/locationtech/rasterframes/expressions/localops/Resample.scala

```diff
@@ -22,53 +22,171 @@
 package org.locationtech.rasterframes.expressions.localops
 
 import geotrellis.raster.Tile
-import geotrellis.raster.resample.NearestNeighbor
+import geotrellis.raster.resample._
+import geotrellis.raster.resample.{Max ⇒ RMax, Min ⇒ RMin, Resample ⇒ GTResample}
 import org.apache.spark.sql.Column
 import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult.{TypeCheckFailure, TypeCheckSuccess}
 import org.apache.spark.sql.catalyst.expressions.codegen.CodegenFallback
-import org.apache.spark.sql.catalyst.expressions.{Expression, ExpressionDescription}
+import org.apache.spark.sql.catalyst.expressions.{Expression, ExpressionDescription, Literal, TernaryExpression}
 import org.apache.spark.sql.functions.lit
-import org.locationtech.rasterframes.expressions.BinaryLocalRasterOp
-import org.locationtech.rasterframes.expressions.DynamicExtractors.tileExtractor
+import org.apache.spark.sql.rf.TileUDT
+import org.apache.spark.sql.types.{DataType, StringType}
+import org.apache.spark.unsafe.types.UTF8String
+import org.locationtech.rasterframes.encoders.CatalystSerializer._
+import org.locationtech.rasterframes.expressions.{fpTile, row}
+import org.locationtech.rasterframes.expressions.DynamicExtractors._
 
-@ExpressionDescription(
-  usage = "_FUNC_(tile, factor) - Resample tile to different size based on scalar factor or tile whose dimension to match. Scalar less than one will downsample tile; greater than one will upsample. Uses nearest-neighbor value.",
-  arguments = """
-  Arguments:
-    * tile - tile
-    * rhs  - scalar or tile to match dimension""",
-  examples = """
-  Examples:
-    > SELECT _FUNC_(tile, 2.0);
-       ...
-    > SELECT _FUNC_(tile1, tile2);
-       ..."""
-)
-case class Resample(left: Expression, right: Expression) extends BinaryLocalRasterOp
+
+abstract class ResampleBase(left: Expression, right: Expression, method: Expression)
+  extends TernaryExpression
   with CodegenFallback {
+
+  import Resample.stringToMethod
+
   override val nodeName: String = "rf_resample"
-  override protected def op(left: Tile, right: Tile): Tile = left.resample(right.cols, right.rows, NearestNeighbor)
-  override protected def op(left: Tile, right: Double): Tile = left.resample((left.cols * right).toInt,
-                                                                             (left.rows * right).toInt, NearestNeighbor)
-  override protected def op(left: Tile, right: Int): Tile = op(left, right.toDouble)
+  override def dataType: DataType = left.dataType
+  override def children: Seq[Expression] = Seq(left, right, method)
+
+  def targetFloatIfNeeded(t: Tile, method: ResampleMethod): Tile =
+     method match {
+      case NearestNeighbor | Mode | RMax | RMin | Sum  ⇒ t
```

Comment:
would like to discuss this in the docs. basically these methods are guaranteed to return an integer type. However you can still get (with sum) some bad behavior like uint8 sum over a large neighborhood resulting in > 255.

Reply:
Oooh, good points. Shoot.  Wish we could detect overflow. Does GeoTrellis have any comments on this?

Reply:
i did not see anything.

Reply:
do you think we should just coerce to float if using Max as it is possible and maybe even likely to overflow??

Reply:
From where I sit, I think coercing to float all the time is OK, but not sure if that would irritate an analyst.